src/checkpoint/traits/user_traits.h

```diff
@@ -0,0 +1,196 @@
+  struct without_helper {
+    using type = typename without_helper<typename Traits::template _without_trait<T>, U...>::type;
+  };
+  template<typename Traits, typename T>
+  struct without_helper<Traits, T> {
+    using type = typename Traits::template _without_trait<T>;
+  };
+}
+
+namespace checkpoint {
+namespace SerializerUserTraits {
+
+  template<typename U, typename V>
+  struct is_base_or_derived
+    : std::disjunction<std::is_base_of<U,V>, std::is_base_of<V,U>> {};
+
+  template<typename U, typename V>
+  struct is_same_template
+    : std::false_type {};
+  template<template <typename...> typename Templ, typename... U, typename... V>
+  struct is_same_template<Templ<U...>, Templ<V...>>
+    : std::true_type {};
+
+  template<typename U, typename V>
+  struct traits_match
+    : std::disjunction<
+#ifdef CHECKPOINT_INHERITED_USER_TRAITS
+        is_base_or_derived<std::decay_t<U>,std::decay_t<V>>,
+#endif
+#ifdef CHECKPOINT_TEMPLATED_USER_TRAITS
+        is_same_template<std::decay_t<U>,std::decay_t<V>>,
+#endif
+        std::is_same<std::decay_t<U>,std::decay_t<V>>
+      > {};
+  template<typename U, typename V>
+  inline constexpr bool traits_match_v = traits_match<U,V>::value;
+
+
+
+  template<typename Trait = NoTrait, typename... Traits>
+  struct TraitHolder {
+  protected:
+    template<typename, typename...>
+    friend struct TraitHolder;
+
+    template<typename T>
+    using has_trait = std::disjunction<traits_match<T, Trait>, traits_match<T, Traits>...>;
+
+  public:
+
+    //Remove one at a time, so we can allow multiple trait copies to be removed independently.
+    template<typename T>
+    using _without_trait = std::conditional_t<
+                            traits_match_v<T, Trait>,
+                            TraitHolder<Traits...>,
+                            typename TraitHolder<Traits...>::template _without_trait<T>::template with_pre<Trait>
+                          >;
+    template<typename... T>
+    using with = TraitHolder<Trait, Traits..., T...>;
+    //To respect ordering. Could be handy in the future -- disambiguating multiple hooks?
+    template<typename... T>
+    using with_pre = TraitHolder<T..., Trait, Traits...>;
+
+    //Flush out any nested traits (TraitHolder<..., TraitHolder<...>, ...>)
+    //Remove any NoTraits, unless that's the only trait.
+    using BaseTraits = typename TraitHolder<Traits...>::BaseTraits::template with_pre<Trait>;
+
+    template<typename... T>
+    using has = std::conjunction<has_trait<T>...>;
+    template<typename... T>
+    using has_any = std::disjunction<has_trait<T>...>;
+
+    template<typename... T>
+    static inline constexpr bool has_v = has<T...>::value;
+    template<typename... T>
+    static inline constexpr bool has_any_v = has_any<T...>::value;
+
+    template<typename T, typename... U>
+    using without = typename without_helper<BaseTraits, T, U...>::type;
+  };
+
+  //Empty specialization w/ shortcuts.
+  template<>
```

Comment:
Having a NoTrait instead of an empty variadic helps simplify the main logic by allowing us to always assume at least one type is in the TraitHolders being passed around